README.md
[![Build Status](https://travis-ci.org/moznion/Perl-PrereqScanner-Lite.svg?branch=master)](https://travis-ci.org/moznion/Perl-PrereqScanner-Lite) [![Coverage Status](https://img.shields.io/coveralls/moznion/Perl-PrereqScanner-Lite/master.svg)](https://coveralls.io/r/moznion/Perl-PrereqScanner-Lite?branch=master)
# NAME

Perl::PrereqScanner::Lite - Lightweight Prereqs Scanner for Perl

# SYNOPSIS

    use Perl::PrereqScanner::Lite;

    my $scanner = Perl::PrereqScanner::Lite->new;
    $scanner->add_extra_scanner('Moose'); # add extra scanner for moose style
    my $modules = $scanner->scan_file('path/to/file');

# DESCRIPTION

Perl::PrereqScanner::Lite is the lightweight prereqs scanner for perl.
This scanner uses [Compiler::Lexer](https://metacpan.org/pod/Compiler::Lexer) as tokenizer, therefore processing speed is really fast.

# METHODS

## new($opt)

Create a scanner instance.

`$opt` must be hash reference. It accepts following keys of hash:

- extra\_scanners

    It specifies extra scanners. This item must be array reference.

    e.g.

        my $scanner = Perl::PrereqScanner::Lite->new(
            extra_scanners => [qw/Moose Version/]
        );

    See also ["add\_extra\_scanner($scanner\_name)"](#add_extra_scanner-scanner_name).

- no\_prereq

    It specifies to use `## no prereq` or not. Please see also ["ADDITIONAL NOTATION"](#additional-notation).

## scan\_file($file\_path)

Scan and figure out prereqs which is instance of `CPAN::Meta::Requirements` by file path.

## scan\_string($string)

Scan and figure out prereqs which is instance of `CPAN::Meta::Requirements` by source code string written in perl.

e.g.

    open my $fh, '<', __FILE__;
    my $string = do { local $/; <$fh> };
    my $modules = $scanner->scan_string($string);

## scan\_module($module\_name)

Scan and figure out prereqs which is instance of `CPAN::Meta::Requirements` by module name.

e.g.

    my $modules = $scanner->scan_module('Perl::PrereqScanner::Lite');

## scan\_tokens($tokens)

Scan and figure out prereqs which is instance of `CPAN::Meta::Requirements` by tokens of [Compiler::Lexer](https://metacpan.org/pod/Compiler::Lexer).

e.g.

    open my $fh, '<', __FILE__;
    my $string = do { local $/; <$fh> };
    my $tokens = Compiler::Lexer->new->tokenize($string);
    my $modules = $scanner->scan_tokens($tokens);

## add\_extra\_scanner($scanner\_name)

Add extra scanner to scan and figure out prereqs. This module loads extra scanner such as `Perl::PrereqScanner::Lite::Scanner::$scanner_name` if specifying scanner name through this method.

If you want to specify an extra scanner from external package without `Perl::PrereqScanner::Lite::` prefix, you can prepend `+` to `$scanner_name`. Like so `+Your::Awesome::Scanner`.

Extra scanners that are default supported are followings;

- [Perl::PrereqScanner::Lite::Scanner::Moose](https://metacpan.org/pod/Perl::PrereqScanner::Lite::Scanner::Moose)
- [Perl::PrereqScanner::Lite::Scanner::Version](https://metacpan.org/pod/Perl::PrereqScanner::Lite::Scanner::Version)

# ADDITIONAL NOTATION

If `no_prereq` is enabled by `new()` (like so: `Perl::PrereqScanner::Lite->new({no_prereq => 1})`),
this module recognize `## no prereq` optional comment. The requiring declaration with this comment on the same line will be ignored as prereq.

For example

    use Foo;
    use Bar; ## no prereq

In this case `Foo` is the prereq, however `Bar` is ignored.

# SPEED COMPARISON

## Plain

                                Rate   Perl::PrereqScanner Perl::PrereqScanner::Lite
    Perl::PrereqScanner       8.57/s                    --                      -97%
    Perl::PrereqScanner::Lite  246/s                 2770%                        --

## With Moose scanner

                                Rate   Perl::PrereqScanner Perl::PrereqScanner::Lite
    Perl::PrereqScanner       9.00/s                    --                      -94%
    Perl::PrereqScanner::Lite  152/s                 1587%                        --

# SEE ALSO

[Perl::PrereqScanner](https://metacpan.org/pod/Perl::PrereqScanner), [Compiler::Lexer](https://metacpan.org/pod/Compiler::Lexer)

# LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

moznion <moznion@gmail.com>


Changes
Revision history for Perl extension Perl-PrereqScanner-Lite

{{$NEXT}}

0.28 2017-11-17T07:02:04Z

    - Fix tests: problem of `@INC` doesn't contain current dir (Fixed #18)

0.27 2017-11-17T02:11:36Z

    - Fix v_string test to work with modern version.pm (@ctriv++)

0.26 2015-05-14T16:15:05Z

    - Fix moose scanner (@charsbar++)

0.25 2015-04-18T12:14:20Z

    - Fix some Moose scanners bugs

0.24 2015-02-28T05:01:13Z

    - Improve plugin mechanism for extra scanners

0.23 2015-02-22T10:59:30Z

    - Fix lexical analysis processing
      (#5, Thank you for reporting @perlancar)

0.22 2015-02-19T02:46:39Z

    - Fix the problem is that it truncates v-string if it has underscore
      (#6, Thank you for reporting @miyagawa)

0.21 2014-09-05T15:22:30Z

    - Fix the bug that mixes the blank in dependencies (Thank you for reporting @jscook)

0.20 2014-07-31T07:59:24Z

    - Bug fix (#2)
    - Upgrade the dependency of Compiler::Lexer

0.19 2014-07-17T01:37:16Z

    - Bug fix (#3)

0.18 2014-04-17T07:00:00Z

    - Support `no` statement to scan prereqs

0.17 2014-04-03T13:19:35Z

    - Fix miss detection
    - Add documentation about speed comparison

0.16 2014-04-03T07:04:04Z

    - Fix scanner for Moose

0.15 2014-04-02T15:38:20Z

    - Support `with` notation for Moose
    - Support `aliased` notation

0.14 2014-03-29T04:44:12Z

    - Change "## no prereq" to be option

0.13 2014-03-29T03:49:54Z

    - Support "## no prereq" comment

0.12 2014-03-21T14:35:16Z

    - Fix the handling of `use constant`

0.11 2014-03-21T13:44:07Z

    - Bug fix

0.10 2014-03-21T09:43:52Z

    - Change data representation format of prereqs
      (Hash -> Instance of CPAN::Meta::Requirements)

0.01 2014-03-21T08:37:04Z

    - original version



t/Util.pm
package t::Util;
use strict;
use warnings;
use utf8;
use CPAN::Meta::Requirements;
use Test::Deep;

use parent qw/Exporter/;

our @EXPORT = qw/slurp get_reqs_hash prereqs_ok/;

sub slurp {
    my ($file_path) = @_;

    open my $fh, '<', $file_path;
    do { local $/; <$fh>; };
}

sub get_reqs_hash {
    my ($req) = @_;

    CPAN::Meta::Requirements->new->add_requirements($req)->as_string_hash;
}

sub prereqs_ok {
    my ($got) = @_;
    cmp_deeply(get_reqs_hash($got), {
        'strict'       => 0,
        'warnings'     => 0,
        'parent'       => 0,
        'base'         => 0,
        'lib'          => 0,
        'constant'     => 0,
        'aliased'      => 0,
        'perl'         => 'v5.8.1',
        'Time::Local'  => 0,
        'Exporter'     => 0,
        'File::Temp'   => '0.1_2',
        'Fcntl'        => 0,
        'FileHandle'   => 0,
        'Env'          => 0,
        'English'      => 0,
        'Carp'         => 0,
        'Cwd'          => 0,
        'Getopt::Long' => 0,
        'Getopt::Std'  => 0,
        'TieHash'      => 0,
        'Text::Tabs'   => 0,
        'PerlIO'       => 0,
        'Opcode'       => 0,
        'Pod::Checker' => 0,
        'Pod::Find'    => 0,
        'JSON'         => 2,
        'Test::More'   => 0,
        'Perl::PrereqScanner'       => 0,
        'Perl::PrereqScanner::Lite' => 0,
        'Locale::MakeText'          => 0,
        'Locale::MakeText::Lexicon' => 0,
        'Moose' => 0,
        'Any::Moose' => 0,
    });
}

1;



t/resources/foo.pl
use if $] < 5.009_005, 'MRO::Compat';
use Test::More tests => 4;

use strict;
use warnings;
use Time::Local qw(timelocal);
use Exporter ();
use JSON 2 qw/encode_json/;
use File::Temp 0.1_1 ();
use File::Temp 0.1_2 ();
use v5.8.1;
use parent ("Fcntl", 'FileHandle');
use parent qw/Env English/;
use parent "PerlIO";
use parent 'Opcode';
use base ("Carp", 'Cwd');
use base qw/Getopt::Long Getopt::Std/;
use base "Pod::Checker";
use base 'Pod::Find';
use lib "$FinBin::Bin/../lib";
use constant FOO => 'BAR';
use aliased 'Perl::PrereqScanner' => 'P::PS';
use aliased "Perl::PrereqScanner::Lite" => 'P::PS::Lite';
require TieHash;
require Text::Tabs;
require "Text/Soundex.pm"; # <= should be ignored

use Locale::MakeText {
    ja => [ Gettext => File::Spec->catdir() ],
};
use Locale::MakeText::Lexicon [
    ja => [ Gettext => File::Spec->catdir() ],
];

use Sys::Syslog; ## no prereq

1;



t/resources/basic.pl
use if $] < 5.009_005, 'MRO::Compat';
use Test::More tests => 4;

use strict;
use warnings;
use Time::Local qw(timelocal);
use Exporter ();
use JSON 2 qw/encode_json/;
use File::Temp 0.1_1 ();
use File::Temp 0.1_2 ();
use v5.8.1;
use parent ("Fcntl", 'FileHandle');
use parent qw/Env English/;
use parent "PerlIO";
use parent 'Opcode';
use base ("Carp", 'Cwd');
use base qw/Getopt::Long Getopt::Std/;
use base "Pod::Checker";
use base 'Pod::Find';
use lib "$FinBin::Bin/../lib";
use constant FOO => 'BAR';
use aliased 'Perl::PrereqScanner' => 'P::PS';
use aliased "Perl::PrereqScanner::Lite" => 'P::PS::Lite';
require TieHash;
require Text::Tabs;
require "Text/Soundex.pm"; # <= should be ignored

use Locale::MakeText {
    ja => [ Gettext => File::Spec->catdir() ],
};
use Locale::MakeText::Lexicon [
    ja => [ Gettext => File::Spec->catdir() ],
];

use Sys::Syslog; ## no prereq

no Moose;
no Any::Moose;

1;



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Perl-PrereqScanner-Lite',
    module_name     => 'Perl::PrereqScanner::Lite',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Perl/PrereqScanner/Lite/Constants.pm
package Perl::PrereqScanner::Lite::Constants;
use strict;
use warnings;
use utf8;
use Compiler::Lexer::Constants;

use parent qw(Exporter);

our @EXPORT = qw(
    REQUIRE_DECL REQUIRED_NAME NAMESPACE_RESOLVER NAMESPACE
    SEMI_COLON USE_DECL USED_NAME REG_LIST REG_EXP LEFT_PAREN
    RIGHT_PAREN STRING RAW_STRING VERSION_STRING INT DOUBLE KEY
    METHOD WHITESPACE COMMENT LEFT_BRACE LEFT_BRACKET BUILTIN_FUNC
    IF_STMT COMMA
);

use constant {
    REQUIRE_DECL       => Compiler::Lexer::TokenType::T_RequireDecl,
    REQUIRED_NAME      => Compiler::Lexer::TokenType::T_RequiredName,
    NAMESPACE_RESOLVER => Compiler::Lexer::TokenType::T_NamespaceResolver,
    NAMESPACE          => Compiler::Lexer::TokenType::T_Namespace,
    SEMI_COLON         => Compiler::Lexer::TokenType::T_SemiColon,
    USE_DECL           => Compiler::Lexer::TokenType::T_UseDecl,
    USED_NAME          => Compiler::Lexer::TokenType::T_UsedName,
    REG_LIST           => Compiler::Lexer::TokenType::T_RegList,
    REG_EXP            => Compiler::Lexer::TokenType::T_RegExp,
    STRING             => Compiler::Lexer::TokenType::T_String,
    RAW_STRING         => Compiler::Lexer::TokenType::T_RawString,
    VERSION_STRING     => Compiler::Lexer::TokenType::T_VersionString,
    INT                => Compiler::Lexer::TokenType::T_Int,
    DOUBLE             => Compiler::Lexer::TokenType::T_Double,
    KEY                => Compiler::Lexer::TokenType::T_Key,
    METHOD             => Compiler::Lexer::TokenType::T_Method,
    WHITESPACE         => Compiler::Lexer::TokenType::T_WhiteSpace,
    COMMENT            => Compiler::Lexer::TokenType::T_Comment,
    IF_STMT            => Compiler::Lexer::TokenType::T_IfStmt,
    COMMA              => Compiler::Lexer::TokenType::T_Comma,

    LEFT_PAREN   => Compiler::Lexer::TokenType::T_LeftParenthesis,
    RIGHT_PAREN  => Compiler::Lexer::TokenType::T_RightParenthesis,
    LEFT_BRACE   => Compiler::Lexer::TokenType::T_LeftBrace,
    LEFT_BRACKET => Compiler::Lexer::TokenType::T_LeftBracket,

    BUILTIN_FUNC => Compiler::Lexer::TokenType::T_BuiltinFunc,
};

1;



lib/Perl/PrereqScanner/Lite/Scanner/Moose.pm
package Perl::PrereqScanner::Lite::Scanner::Moose;
use strict;
use warnings;
use utf8;
use Perl::PrereqScanner::Lite::Constants;

sub scan {
    my ($class, $c, $token, $token_type) = @_;

    my $token_data = $token->data;
    if ($token_type == KEY && ($token_data eq 'extends' || $token_data eq 'with')) {
        $c->{is_in_moose_inherited} = 1;
        return 1;
    }

    if ($c->{is_in_moose_inherited}) {
        # to skip content which is is curly bracket -> { ... }
        {
            if ($token_type == LEFT_BRACE) {
                $c->{is_in_moose_role_def} = 1;
                return 1;
            }

            if ($token_type == RIGHT_PAREN) {
                $c->{is_in_moose_role_def} = 0;
                return 1;
            }

            if ($c->{is_in_moose_role_def}) {
                return 1;
            }
        }

        # For qw() notation
        # e.g.
        #   extends qw/Foo Bar/;
        #   with qw/Foo Bar/;
        if ($token_type == REG_LIST) {
            $c->{is_in_moose_inherited_reglist} = 1;
            return 1;
        }
        if ($c->{is_in_moose_inherited_reglist} && !$c->{does_exist_moose_garbage}) {
            if ($token_type == REG_EXP) {
                for my $_module_name (split /\s+/, $token_data) {
                    $c->add_minimum($_module_name => 0);
                }
                $c->{is_in_moose_inherited_reglist} = 0;
            }
            return 1;
        }

        # For simply list
        # e.g.
        #   extends ('Foo', 'Bar');
        #   with ('Foo', 'Bar');
        if ($token_type == LEFT_PAREN) {
            $c->{is_in_moose_inherited_list} = 1;
            return 1;
        }
        if ($token_type == RIGHT_PAREN) {
            $c->{is_in_moose_inherited_list} = 0;
            return 1;
        }
        if ($c->{is_in_moose_inherited_list}) {
            if (($token_type == STRING || $token_type == RAW_STRING) && !$c->{does_exist_moose_garbage}) {
                $c->add_minimum($token_data => 0);
            }
            return 1;
        }

        # For string
        # e.g.
        #   extends "Foo"
        #   with "Foo"
        if ((($token_type == STRING || $token_type == RAW_STRING)) && !$c->{does_exist_moose_garbage}) {
            $c->add_minimum($token_data => 0);
            return 1;
        }

        # End of extends or with
        if ($token_type == SEMI_COLON) {
            $c->{is_in_moose_inherited}         = 0;
            $c->{is_in_moose_inherited_reglist} = 0;
            $c->{is_in_moose_inherited_list}    = 0;
            $c->{does_exist_moose_garbage}      = 0;
            return 1;
        }

        # For
        #   extends 'Class1', 'Class2';
        if ($token_type != COMMA) {
            $c->{does_exist_moose_garbage} = 1;
        }

        return 1;
    }

    return;
}

1;

=encoding utf-8

=head1 NAME

Perl::PrereqScanner::Lite::Scanner::Moose - Extra Perl::PrereqScanner::Lite Scanner for Moose Family

=head1 SYNOPSIS

    use Perl::PrereqScanner::Lite;

    my $scanner = Perl::PrereqScanner::Lite->new;
    $scanner->add_extra_scanner('Moose');

=head1 DESCRIPTION

Perl::PrereqScanner::Lite::Scanner::Moose is the extra scanner for Perl::PrereqScanner::Lite. This scanner supports C<extends> and C<with> notation for Moose family.

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut



lib/Perl/PrereqScanner/Lite/Scanner/Version.pm
package Perl::PrereqScanner::Lite::Scanner::Version;
use strict;
use warnings;
use utf8;
use Perl::PrereqScanner::Lite::Constants;

sub scan {
    my ($class, $c, $token, $token_type) = @_;

    if ($token_type == KEY || $token_type == NAMESPACE || $token_type == NAMESPACE_RESOLVER) {
        $c->{not_decl_module_name} .= $token->{data};
        return 1;
    }

    if ($token_type == METHOD && $token->{data} eq 'VERSION') {
        $c->{is_version_decl} = 1;
        return 1;
    }

    if ($c->{is_version_decl} && $token_type == INT || $token_type == DOUBLE || $token_type == VERSION_STRING) {
        if ($c->{module_reqs}->{requirements}->{$c->{not_decl_module_name}}) {
            $c->_add_minimum($c->{not_decl_module_name} => $token->{data});
        }
        $c->{is_version_decl} = 0;
        $c->{not_decl_module_name} = '';
        return 1;
    }

    if ($token_type == SEMI_COLON) {
        $c->{is_version_decl} = 0;
        $c->{not_decl_module_name} = '';
        return 1;
    }
}

1;

=encoding utf-8

=head1 NAME

Perl::PrereqScanner::Lite::Scanner::Version - Extra Perl::PrereqScanner::Lite Scanner for VERSION method

=head1 SYNOPSIS

    use Perl::PrereqScanner::Lite;

    my $scanner = Perl::PrereqScanner::Lite->new;
    $scanner->add_extra_scanner('Version');

=head1 DESCRIPTION

Perl::PrereqScanner::Lite::Scanner::Version is the extra scanner for Perl::PrereqScanner::Lite.
This scanner supports C<VERSION> method. It retrieves version from the argument of C<VERSION>.

For example,

    require Foo::Bar;
    Foo::Bar->VERSION(1.00);

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut



lib/Perl/PrereqScanner/Lite.pm
package Perl::PrereqScanner::Lite;
use 5.008005;
use strict;
use warnings;
use Carp ();
use Compiler::Lexer;
use CPAN::Meta::Requirements;
use Perl::PrereqScanner::Lite::Constants;

our $VERSION = "0.27";

sub new {
    my ($class, $opt) = @_;

    my $lexer;
    if ($opt->{no_prereq}) {
        $lexer = Compiler::Lexer->new({verbose => 1}),
    }
    else {
        $lexer = Compiler::Lexer->new(),
    }

    my $extra_scanners = [];
    if (my $scanner_names = $opt->{extra_scanners}) {
        if (ref $scanner_names eq 'ARRAY') {
            for my $scanner_name (@$scanner_names) {
                my $extra_scanner;
                if (substr($scanner_name, 0, 1) eq '+') {
                    $extra_scanner = substr $scanner_name, 1;
                }
                else {
                    $extra_scanner = "Perl::PrereqScanner::Lite::Scanner::$scanner_name";
                }

                eval "require $extra_scanner"; ## no critic
                push @$extra_scanners, $extra_scanner;
            }
        } else {
            Carp::croak "'extra_scanners' option must be array reference";
        }
    }

    bless {
        lexer          => $lexer,
        extra_scanners => $extra_scanners,
        module_reqs    => CPAN::Meta::Requirements->new,
    }, $class;
}

sub add_extra_scanner {
    my ($self, $scanner_name) = @_;

    my $extra_scanner;
    if (substr($scanner_name, 0, 1) eq '+') {
        $extra_scanner = substr $scanner_name, 1;
    }
    else {
        $extra_scanner = "Perl::PrereqScanner::Lite::Scanner::$scanner_name";
    }

    eval "require $extra_scanner"; ## no critic
    push @{$self->{extra_scanners}}, $extra_scanner;
}

sub scan_string {
    my ($self, $string) = @_;

    my $tokens = $self->{lexer}->tokenize($string);
    $self->_scan($tokens);
}

sub scan_file {
    my ($self, $file) = @_;

    open my $fh, '<', $file or die "Cannot open file: $file";
    my $script = do { local $/; <$fh>; };

    $self->scan_string($script);
}

sub scan_tokens {
    my ($self, $tokens) = @_;
    $self->_scan($tokens);
}

sub scan_module {
    my ($self, $module) = @_;

    require Module::Path;

    if (defined(my $path = Module::Path::module_path($module))) {
        return $self->scan_file($path);
    }
}

sub _scan {
    my ($self, $tokens) = @_;

    my $module_name    = '';
    my $module_version = 0;

    my $not_decl_module_name = '';

    my $is_in_reglist   = 0;
    my $is_in_usedecl   = 0;
    my $is_in_reqdecl   = 0;
    my $is_inherited    = 0;
    my $is_in_list      = 0;
    my $is_version_decl = 0;
    my $is_aliased      = 0;
    my $is_prev_version = 0;
    my $is_prev_module_name = 0;

    my $does_garbage_exist = 0;
    my $does_use_lib_or_constant = 0;

    my $latest_prereq = '';

    TOP:
    for (my $i = 0; my $token = $tokens->[$i]; $i++) {
        my $token_type = $token->{type};

        # For require statement
        if ($token_type == REQUIRE_DECL || ($token_type == BUILTIN_FUNC && $token->{data} eq 'no')) {
            $is_in_reqdecl = 1;
            next;
        }
        if ($is_in_reqdecl) {
            # e.g.
            #   require Foo;
            if ($token_type == REQUIRED_NAME || $token_type == KEY) {
                $latest_prereq = $self->add_minimum($token->{data} => 0);

                $is_in_reqdecl = 0;
                next;
            }

            # e.g.
            #   require Foo::Bar;
            if ($token_type == NAMESPACE || $token_type == NAMESPACE_RESOLVER) {
                $module_name .= $token->{data};
                next;
            }

            # End of declare of require statement
            if ($token_type == SEMI_COLON) {
                if ($module_name) {
                    $latest_prereq = $self->add_minimum($module_name => 0);
                }

                $module_name   = '';
                $is_in_reqdecl = 0;
                next;
            }

            next;
        }

        # For use statement
        if ($token_type == USE_DECL) {
            $is_in_usedecl = 1;
            next;
        }
        if ($is_in_usedecl) {
            # e.g.
            #   use Foo;
            #   use parent qw/Foo/;
            #
            if ($token_type == USED_NAME || $token_type == IF_STMT) {
                # XXX                       ~~~~~~~~~~~~~~~~~~~~~~
                # Workaround for `use if` statement
                # It is a matter of Compiler::Lexer (maybe).
                #
                #   use if $] < 5.009_005, 'MRO::Compat';

                $module_name = $token->{data};

                if ($module_name eq 'lib' || $module_name eq 'constant') {
                    $latest_prereq = $self->add_minimum($module_name, 0);
                    $does_use_lib_or_constant = 1;
                }
                elsif ($module_name =~ /(?:base|parent)/) {
                    $is_inherited = 1;
                }
                elsif ($module_name =~ 'aliased') {
                    $is_aliased = 1;
                }

                $is_prev_module_name = 1;
                next;
            }

            # End of declare of use statement
            if ($token_type == SEMI_COLON || $token_type == LEFT_BRACE || $token_type == LEFT_BRACKET) {
                if ($module_name && !$does_use_lib_or_constant) {
                    $latest_prereq = $self->add_minimum($module_name => $module_version);
                }

                $module_name    = '';
                $module_version = 0;
                $is_in_reglist  = 0;
                $is_inherited   = 0;
                $is_in_list     = 0;
                $is_in_usedecl  = 0;
                $is_aliased     = 0;
                $does_garbage_exist  = 0;
                $is_prev_module_name = 0;
                $does_use_lib_or_constant = 0;

                next;
            }

            # e.g.
            #   use Foo::Bar;
            if ($token_type == NAMESPACE || $token_type == NAMESPACE_RESOLVER) {
                $module_name .= $token->{data};
                $is_prev_module_name = 1;
                next;
            }

            # Section for parent/base
            if ($is_inherited) {
                # For qw() notation
                # e.g.
                #   use parent qw/Foo Bar/;
                if ($token_type == REG_LIST) {
                    $is_in_reglist = 1;
                }
                elsif ($is_in_reglist) {
                    if ($token_type == REG_EXP) {
                        for my $_module_name (split /\s+/, $token->{data}) {
                            $latest_prereq = $self->add_minimum($_module_name => 0);
                        }
                        $is_in_reglist = 0;
                    }
                }

                # For simply list
                # e.g.
                #   use parent ('Foo' 'Bar');
                elsif ($token_type == LEFT_PAREN) {
                    $is_in_list = 1;
                }
                elsif ($token_type == RIGHT_PAREN) {
                    $is_in_list = 0;
                }
                elsif ($is_in_list) {
                    if ($token_type == STRING || $token_type == RAW_STRING) {
                        $latest_prereq = $self->add_minimum($token->{data} => 0);
                    }
                }

                # For string
                # e.g.
                #   use parent "Foo"
                elsif ($token_type == STRING || $token_type == RAW_STRING) {
                    $latest_prereq = $self->add_minimum($token->{data} => 0);
                }

                $is_prev_module_name = 0;
                next;
            }

            if ($token_type == DOUBLE || $token_type == INT || $token_type == VERSION_STRING) {
                if (!$module_name) {
                    if (!$does_garbage_exist) {
                        # For perl version
                        # e.g.
                        #   use 5.012;
                        my $perl_version = $token->{data};
                        $latest_prereq = $self->add_minimum('perl' => $perl_version);
                        $is_in_usedecl = 0;
                    }
                }
                elsif($is_prev_module_name) {
                    # For module version
                    # e.g.
                    #   use Foo::Bar 0.0.1;'
                    #   use Foo::Bar v0.0.1;
                    #   use Foo::Bar 0.0_1;
                    $module_version = $token->{data};
                }

                $is_prev_module_name = 0;
                $is_prev_version = 1;
                next;
            }

            if ($is_aliased) {
                if ($token_type == STRING || $token_type == RAW_STRING) {
                    $latest_prereq = $self->add_minimum($token->{data} => 0);
                    $is_aliased = 0;
                }
                next;
            }

            if (($is_prev_module_name || $is_prev_version) && $token_type == LEFT_PAREN) {
                my $left_paren_num = 1;
                for ($i++; $token = $tokens->[$i]; $i++) { # skip content that is surrounded by parens
                    $token_type = $token->{type};

                    if ($token_type == LEFT_PAREN) {
                        $left_paren_num++;
                    }
                    elsif ($token_type == RIGHT_PAREN) {
                        last if --$left_paren_num <= 0;
                    }
                }
                next;
            }

            if ($token_type != WHITESPACE) {
                $does_garbage_exist  = 1;
                $is_prev_module_name = 0;
                $is_prev_version = 0;
            }
            next;
        }

        for my $extra_scanner (@{$self->{extra_scanners}}) {
            if ($extra_scanner->scan($self, $token, $token_type)) {
                next TOP;
            }
        }

        if ($token_type == COMMENT && $token->{data} =~ /\A##\s*no prereq\Z/) {
            $self->{module_reqs}->clear_requirement($latest_prereq);
            next;
        }
    }

    return $self->{module_reqs};
}

sub add_minimum {
    my ($self, $module_name, $module_version) = @_;

    if ($module_name) {
        $self->{module_reqs}->add_minimum($module_name => $module_version);
    }

    return $module_name;
}

1;
__END__

=encoding utf-8

=for stopwords prepend reimplement tokenizer

=head1 NAME

Perl::PrereqScanner::Lite - Lightweight Prereqs Scanner for Perl

=head1 SYNOPSIS

    use Perl::PrereqScanner::Lite;

    my $scanner = Perl::PrereqScanner::Lite->new;
    $scanner->add_extra_scanner('Moose'); # add extra scanner for moose style
    my $modules = $scanner->scan_file('path/to/file');

=head1 DESCRIPTION

Perl::PrereqScanner::Lite is the lightweight prereqs scanner for perl.
This scanner uses L<Compiler::Lexer> as tokenizer, therefore processing speed is really fast.

=head1 METHODS

=head2 new($opt)

Create a scanner instance.

C<$opt> must be hash reference. It accepts following keys of hash:

=over 4

=item * extra_scanners

It specifies extra scanners. This item must be array reference.

e.g.

    my $scanner = Perl::PrereqScanner::Lite->new(
        extra_scanners => [qw/Moose Version/]
    );

See also L</add_extra_scanner($scanner_name)>.

=item * no_prereq

It specifies to use C<## no prereq> or not. Please see also L</ADDITIONAL NOTATION>.

=back

=head2 scan_file($file_path)

Scan and figure out prereqs which is instance of C<CPAN::Meta::Requirements> by file path.

=head2 scan_string($string)

Scan and figure out prereqs which is instance of C<CPAN::Meta::Requirements> by source code string written in perl.

e.g.

    open my $fh, '<', __FILE__;
    my $string = do { local $/; <$fh> };
    my $modules = $scanner->scan_string($string);

=head2 scan_module($module_name)

Scan and figure out prereqs which is instance of C<CPAN::Meta::Requirements> by module name.

e.g.

    my $modules = $scanner->scan_module('Perl::PrereqScanner::Lite');

=head2 scan_tokens($tokens)

Scan and figure out prereqs which is instance of C<CPAN::Meta::Requirements> by tokens of L<Compiler::Lexer>.

e.g.

    open my $fh, '<', __FILE__;
    my $string = do { local $/; <$fh> };
    my $tokens = Compiler::Lexer->new->tokenize($string);
    my $modules = $scanner->scan_tokens($tokens);

=head2 add_extra_scanner($scanner_name)

Add extra scanner to scan and figure out prereqs. This module loads extra scanner such as C<Perl::PrereqScanner::Lite::Scanner::$scanner_name> if specifying scanner name through this method.

If you want to specify an extra scanner from external package without C<Perl::PrereqScanner::Lite::> prefix, you can prepend C<+> to C<$scanner_name>. Like so C<+Your::Awesome::Scanner>.

Extra scanners that are default supported are followings;

=over 8

=item * L<Perl::PrereqScanner::Lite::Scanner::Moose>

=item * L<Perl::PrereqScanner::Lite::Scanner::Version>

=back

=head1 ADDITIONAL NOTATION

If C<no_prereq> is enabled by C<new()> (like so: C<Perl::PrereqScanner::Lite-E<gt>new({no_prereq =E<gt> 1})>),
this module recognize C<## no prereq> optional comment. The requiring declaration with this comment on the same line will be ignored as prereq.

For example

    use Foo;
    use Bar; ## no prereq

In this case C<Foo> is the prereq, however C<Bar> is ignored.

=head1 SPEED COMPARISON

=head2 Plain

                                Rate   Perl::PrereqScanner Perl::PrereqScanner::Lite
    Perl::PrereqScanner       8.57/s                    --                      -97%
    Perl::PrereqScanner::Lite  246/s                 2770%                        --

=head2 With Moose scanner

                                Rate   Perl::PrereqScanner Perl::PrereqScanner::Lite
    Perl::PrereqScanner       9.00/s                    --                      -94%
    Perl::PrereqScanner::Lite  152/s                 1587%                        --

=head1 NOTES

This is a quotation from L<https://github.com/moznion/Perl-PrereqScanner-Lite/issues/13>.

=begin text

The interface of an this module object suggests every scan_* call is not affected by any other, yet the code is storing the requirements in that object. This is quite surprising.

I'd suggest that either it must change to be more functional-style, or this behavior should be clearly documented.

=end text

Yes, it's true. This design is so ugly and not smart.
So I have to redesign and reimplement this module, and I have some plans.

If you have a mind to expand this module by implementing external scanner,
please be careful.
Every C<scan_*> calls must not affect to any others through the
singleton of this module (called it C<$c> in L<https://github.com/moznion/Perl-PrereqScanner-Lite/blob/c03638b2e2a39d92f4d7df360af5a6be65dc417a/lib/Perl/PrereqScanner/Lite/Scanner/Moose.pm#L8>).

=head1 SEE ALSO

L<Perl::PrereqScanner>, L<Compiler::Lexer>

=head1 LICENSE

Copyright (C) moznion.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

moznion E<lt>moznion@gmail.comE<gt>

=cut

